Python package

dafyddstephenson opened the pull request:
Hi all,
Here is the initial implementation of the C-Star python package. It can be found in the `cstar_ocean` subdir.

- The readme/setup instructions is/are at [`cstar_ocean/README.md`](https://github.com/dafyddstephenson/C-Star/blob/python_package/cstar_ocean/README.md)
- Examples of use can be found at:
  -  [`cstar_ocean/cstar_example_notebook.ipynb`](https://github.com/dafyddstephenson/C-Star/blob/python_package/cstar_ocean/examples/cstar_example_notebook.ipynb) (more detailed/hand-holding)
  - [`cstar_ocean/cstar_example.py`](https://github.com/dafyddstephenson/C-Star/blob/python_package/cstar_ocean/examples/cstar_example_notebook.ipynb) (simpler, for use in an interactive session by somebody who is exploring the code itself)

There are a few `TODO`s,`QUESTION`s, and `FIXME`s in the code that would be good to discuss here.
Looking forward to discussing!
Dafydd

TomNicholas commented:
This looks great @dafyddstephenson !

This is such a massive pull request I'm not quite sure how to go about reviewing it. I think what I will do is make various comments, but also raise new issues to refer to anything that's bigger than a small code change. That way this can be merged quickly and we can have a solid foundation to develop from.

TomNicholas commented on cstar_ocean/README.md:
```diff
@@ -0,0 +1,86 @@
+
+It may also be worth including `ipython`,`xarray`,`numpy`, `matplotlib` and `jupyter` in the environment to work with any output, though these are not needed to run C-Star.
+
+### ii. Add C-Star to your path
+Use e.g. `sys.path.append(/Users/myname/Code/C-Star/cstar_ocean`) in a python session.
```
We shouldn't need to manually do this just to import a python package. We might need to do it for other reasons - why exactly are we modifying the system `$PATH`?

dafyddstephenson replied:
Hmm, this was a bit of an afterthought as I'd been predominantly either working in the `cstar_ocean` directory itself where the package was visible to python or working with a locally installed version built with conda. 

However, the `examples` directory is deliberately located such that it is unable to see the package without being told where to look for it (to avoid conflicts between installing with conda vs just importing it locally).

Is there a more established way of importing a package that isn't visible to the interpreter?

TomNicholas replied:
If you want to do local development wouldn't you just do `pip install -e .` where `.` is the root directory of the python package? That doesn't require any modifications to the path. And that's separate from how it's all packaged up for conda.

TomNicholas commented on cstar_ocean/README.md:
```diff
@@ -0,0 +1,86 @@
+### ii. Add C-Star to your path
+Use e.g. `sys.path.append(/Users/myname/Code/C-Star/cstar_ocean`) in a python session.
+
+---
+
+## Option 2: Install with conda
+While C-Star isn't yet directly available through conda's usual channels, it can be built and installed locally with conda
+### i. Build the package
+In your `base` conda environment, run `conda-build conda_recipe/` In the `C-Star/cstar_ocean` directory. **If this works, skip to step 2.**
+
+If it did not work, note that you may first have to run `conda install conda-build`
+in the base environment.
+
+If you are on a HPC system, writing to the base environment may not be allowed. If you have this issue, it is recommended to install conda yourself, rather than using Linux environment modules.
+
+To install your own copy of conda on a HPC, first remove any reference to conda in your shell's rc file (e.g. `~/.bashrc`) and use `module restore` to go back to the default module set.
+Then use (noting that step 3 will manipulate your shell's rc file):
+- `wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh`
+- `bash Miniconda3-latest-Linux-x86_64.sh -b -p $HOME/miniconda3`
+- `$HOME/miniconda3/bin/conda init`
+
+You can then try to install `conda-build` in the base environment again.
+
+### ii. Install the package
+Once the package has been built, you can install it using `conda install --use-local cstar_ocean` from any location. It is recommended that you use a clean environment.
```
This whole local conda thing seems very janky, and I think all conda-related stuff (including this documentation) should be removed from this PR and submitted as a later PR.

dafyddstephenson replied:
done in `96f019f55204fb4fd12353dff17a000fda931d1d`. Will update the README shortly.

TomNicholas commented on cstar_ocean/conda_recipe/meta.yaml:
```diff
@@ -0,0 +1,31 @@
+channels:
+  - conda-forge
+  - defaults
```
I'm still very unclear as to why this should live in this repo as opposed to as a PR to conda-feedstocks. Again I think this file should be deferred to a later PR.

dafyddstephenson replied:
done in `96f019f55204fb4fd12353dff17a000fda931d1d`

TomNicholas commented on cstar_ocean/cstar_ocean/__init__.py:
```diff
@@ -0,0 +1,33 @@
+    InitialConditions,
+)
+
+from .cstar_component import Component, ROMSComponent, MARBLComponent
+from .cstar_case import Case
```
We should run the linting tool `absolufy-imports` to make these absolute paths, and do that automatically via `pre-commit`.

dafyddstephenson replied:
done in `be0ceca8025017c04fbd44a4c8acd08320bc8ac5`

TomNicholas commented on cstar_ocean/cstar_ocean/__init__.py:
```diff
@@ -0,0 +1,33 @@
+################################################################################
+# Build module environment at import time
+# NOTE: need to set ROMS_ROOT,MARBL_ROOT,CSTAR_ROOT,CSTAR_SYSTEM, and maybe modify PATH on conda install
+from .cstar_base_model import BaseModel, ROMSBaseModel, MARBLBaseModel
+from .cstar_additional_code import AdditionalCode
+from .cstar_input_dataset import (
```
the `cstar_*` prefixes here are unnecessary, we should change the names of the modules to be short and simple, e.g. `cstar.base`, `cstar.inputs`.

dafyddstephenson replied:
Removed the prefixes for now, except for "case" (`305c3234c49a67932c020b91b5f466c16c4690f6`) but think the module renaming beyond that may require some internal discussion.

TomNicholas replied:
Raised #27 to track discussions on this

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_additional_code.py:
```diff
@@ -0,0 +1,187 @@
+from .cstar_base_model import BaseModel
+
+
+class AdditionalCode:
```
Could we perhaps divide up this PR by separating all `AdditionalCode` functionality into a later PR? I would have thought that the absolute MVP could just assume that there is not additional code.

dafyddstephenson replied:
I feel this should be cleared up some more in the examples but the `AdditionalCode` object is a pretty essential part of a blueprint: it's what separates e.g. pure ROMS from a particular configuration of ROMS. Files like `cppdefs.opt` tell the model whether this configuration uses tidal forcing or not, what sort of physics schemes to use, etc. Other files tell the model the size of the domain, the start date, etc. These are strictly separate from the base ROMS code (described by the `BaseModel` object). It's less important with MARBL, which has a default configuration that is fairly universal. Hope that clears things up?

TomNicholas replied:
Okay that wasn't super clear to me. Conceptually in my head there are choices of inputs / forcings, and then there is the possibility that the user further modified the source code to do something unusual. "Additional Code" implied the latter.

Sounds like we need it in this PR, but should think more about how best to separate out these concepts (and naming things is hard!)

dafyddstephenson commented:
Yes, I realise I should probably have been issuing these incrementally à la roms-tools! 😬 apologies.

Even so there may be some structural things worth overhauling or implementing differently. For instance, the way the environment is managed internally I know was previously flagged as worth addressing ( `cstar_environment.py` and `cstar_local_config.py`, the latter of which is machine-specific and generated by the package when external codebases are installed). 

Are you suggesting it's worth merging this in sooner and then addressing those sorts of things down the line?

Cheers!

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_additional_code.py:
```diff
@@ -0,0 +1,187 @@
+
+    def __init__(
+        self,
+        base_model: BaseModel,
+        source_repo: str,
+        checkout_target: str,
+        source_mods: Optional[Union[str, List[str]]] = None,
+        namelists: Optional[Union[str, List[str]]] = None,
```
Yay for type hints! In python 3.10+ you can use the `|` operator instead of `Union`

https://peps.python.org/pep-0604/

dafyddstephenson replied:
done in [`7b21a14ba79b84ad8b4091afa7dd787ed9194dff`](https://github.com/CWorthy-ocean/C-Star/commit/7b21a14ba79b84ad8b4091afa7dd787ed9194dff)

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_additional_code.py:
```diff
@@ -0,0 +1,187 @@
+
+        """
+
+        # TODO:  Type check here
```
There are python libraries to help with this kind of thing, e.g. pydantic

dafyddstephenson replied:
> Pydantic Example
>`...`
> `from pydantic import BaseModel`

That's fun.

Shall we add this as an issue or do you think it should be handled before the merge?

TomNicholas replied:
I think that can be a separate issue. You (and I) will have to leave more about how pydantic works.

TomNicholas replied:
Raised #26 to track this

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_additional_code.py:
```diff
@@ -0,0 +1,187 @@
+        local_path: str
+            The local path (typically `Case.caseroot`) where the additional code will be curated
+        """
+        with tempfile.TemporaryDirectory() as tmp_dir:
```
I didn't actually know about this `tempfile` module, nice!

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_additional_code.py:
```diff
@@ -0,0 +1,187 @@
+            tgt_dir = local_path + "/" + file_type + "/" + self.base_model.name
+            for f in file_list:
+                tgt_file_path = tgt_dir + "/" + os.path.basename(f)
+                if not os.path.exists(tgt_file_path):
```
There are a lot of places here that would be neater if we used [pathlib](https://docs.python.org/3/library/pathlib.html). That would also help us if we later want to generalize to using [cloudpathlib](https://cloudpathlib.drivendata.org/stable/).

dafyddstephenson replied:
Good idea, seems like a pretty powerful library - issue or tackle now?

TomNicholas replied:
Yeah it's a famously cool part of the python standard library. I often use it as an example when trying to teach people what object-oriented programming is.

I think you can leave this change for later though - just raise an issue.

TomNicholas replied:
Opened https://github.com/CWorthy-ocean/C-Star/issues/27 to track this.

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_base_model.py:
```diff
@@ -0,0 +1,443 @@
+            base_str += " (default)"
+
+        base_str += f"\nlocal_config_status={self.local_config_status} "
+        match self.local_config_status:
+            case 0:
+                base_str += f"(Environment variable {self.expected_env_var} is present, points to the correct repository remote, and is checked out at the correct hash)"
+            case 1:
+                base_str += f"(Environment variable {self.expected_env_var} is present but does not point to the correct repository remote [unresolvable])"
+            case 2:
+                base_str += f"(Environment variable {self.expected_env_var} is present, points to the correct repository remote, but is checked out at the wrong hash)"
+            case 3:
+                base_str += f"(Environment variable {self.expected_env_var} is not present and it is assumed the base model is not installed locally)"
```
This is very interesting

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_base_model.py:
```diff
@@ -0,0 +1,443 @@
+    def expected_env_var(self):
+        """environment variable associated with the base model, e.g. MARBL_ROOT"""
+
+    @abstractmethod
+    def _base_model_adjustments(self):
+        """
+        Perform any C-Star specific adjustments to the base model that would
+        be needed after a clean checkout.
+        """
```
Is this an optional method? Or just not supposed to be public-facing?

dafyddstephenson replied:
It's not really supposed to exist... It's a bit of a bodge tbh, purely due to the ROMS Makefiles issue. Because we have to substitute in our own alternatives to the default Makefiles in ROMS these end up bundled with C-Star (`additional_files`) and this function substitutes them in.

TomNicholas replied:
Okay, sounds like this is an implementation detail. We should treat it as private with the aim to hopefully one day clean up the ROMS build steps enough that this kind of bodge is not longer needed, and can be quietly removed.

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_base_model.py:
```diff
@@ -0,0 +1,443 @@
+                    else:
+                        print("invalid selection; enter 'y','n',or 'custom'")
+
+    @abstractmethod
+    def get(self, target):
+        """clone the basemodel code to your local machine"""
```
These method signatures can also be type-hinted

dafyddstephenson replied:
fixed in [055943542194547112cff6be7ec8b69035d86a3b](https://github.com/CWorthy-ocean/C-Star/commit/055943542194547112cff6be7ec8b69035d86a3b)

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_case.py:
```diff
@@ -0,0 +1,550 @@
+        The name of this case
+    caseroot: str
+        The local directory in which this case will be set up
+    is_from_blueprint: bool
+        Whether this Case was instantiated from a blueprint yaml file
```
Why would it matter how the `Case` was created?

dafyddstephenson replied:
Downstream, there should be some indication of whether a Case is an "approved" blueprint-sourced one or something the user has cooked up. While this isn't strictly that, I think it's good to bake something similar in earlier and keep it in mind.

TomNicholas replied:
We should discuss this in more detail later. I see what you mean, but I also think this could lead to leaky abstractions.

TomNicholas replied:
More thoughts and further discussion in https://github.com/CWorthy-ocean/C-Star/issues/28

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_case.py:
```diff
@@ -0,0 +1,550 @@
+    is_from_blueprint: bool
+        Whether this Case was instantiated from a blueprint yaml file
+
+    Methods
+    -------
+    from_blueprint(blueprint,caseroot)
+        Instantiate a Case from a "blueprint" yaml file
+    persist(filename)
+        Create a "blueprint" yaml file for this Case object
+    setup()
+        Fetch all code and files necessary to run this case in the local caseroot folder
+    build()
+        Compile any case-specific code on this machine
+    pre_run()
+        Execute any pre-processing actions necessary to run this case
+    run()
+        Run this case
+    post_run()
+        Execute any post-processing actions associated with this case
```
This part of the docstring should be redundant with the docstrings of the individual methods, once we have auto-generated API documentation.

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_case.py:
```diff
@@ -0,0 +1,550 @@
+                        else None
+                    )
+
+                    # run_scripts  = [f for f in additional_code_info['scripts']['run']] \
+                    #        if ('scripts'    in additional_code_info.keys()) and \
+                    #           ('run'        in additional_code_info['scripts'].keys()) else None
+
+                    # proc_scripts = [f for f in additional_code_info['scripts']['processing']] \
+                    #        if ('scripts'    in additional_code_info.keys()) and \
+                    #           ('processing' in additional_code_info['scripts'].keys()) else None
```
we don't want to be merging any commented-out code

dafyddstephenson replied:
fixed in [055943542194547112cff6be7ec8b69035d86a3b](https://github.com/CWorthy-ocean/C-Star/commit/055943542194547112cff6be7ec8b69035d86a3b)

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_case.py:
```diff
@@ -0,0 +1,550 @@
+                input_datasets = [
+                    input_datasets,
+                ]
+            if isinstance(input_datasets, list):
+                input_dataset_info: dict = {}
+                for ind in input_datasets:
+                    if isinstance(ind, ModelGrid):
+                        dct_key = "model_grid"
+                    elif isinstance(ind, InitialConditions):
+                        dct_key = "initial_conditions"
+                    elif isinstance(ind, TidalForcing):
+                        dct_key = "tidal_forcing"
+                    elif isinstance(ind, BoundaryForcing):
+                        dct_key = "boundary_forcing"
+                    elif isinstance(ind, SurfaceForcing):
+                        dct_key = "surface_forcing"
+                    if dct_key not in input_dataset_info.keys():
+                        input_dataset_info[dct_key] = {}
+                    if "files" not in input_dataset_info[dct_key].keys():
+                        input_dataset_info[dct_key]["files"] = []
+                    input_dataset_info[dct_key]["files"].append(
+                        {"source": ind.source, "file_hash": ind.file_hash}
+                    )
```
I would think we could add methods to these classes them selves and call the same method on each of them.

TomNicholas commented on cstar_ocean/cstar_ocean/cstar_environment.py:
```diff
@@ -0,0 +1,144 @@
+    if isinstance(spec.submodule_search_locations, list):
+        _CSTAR_ROOT: str = spec.submodule_search_locations[0]
+else:
+    raise ImportError
```
```suggestion
    raise ImportError("some explanatory message")
```
if you don't do this then you haven't actually instantiated the error class

dafyddstephenson replied:
Fixed in [3e12e874b08df27f7fb148c7d9492c2af52d5859](https://github.com/CWorthy-ocean/C-Star/commit/3e12e874b08df27f7fb148c7d9492c2af52d5859)

TomNicholas commented on cstar_ocean/cstar_ocean/utils.py:
```diff
@@ -0,0 +1,149 @@
+
+def _get_repo_remote(local_root):
+    """Take a local repository path string (local_root) and return as a string the remote URL"""
+    return subprocess.run(
+        f"git -C {local_root} remote get-url origin",
+        shell=True,
+        capture_output=True,
+        text=True,
+    ).stdout.strip()
```
I bet there is some python library for handling git commands that would do this for us

dafyddstephenson replied:
I couldn't find anything, but I imagine you'd be more successful at searching! CESM uses [manic](https://github.com/ESMCI/manage_externals/releases/tag/manic-1.2.3)  but it seems to not be traditionally distributed, and any repo I've seen calling for its use (including early versions of C-Star, pre-python) just includes the manic source code. I abandoned manic as it was a lot of bloat for what was effectively in our case just a wrapper for `git clone` and `get checkout`.

TomNicholas replied:
I thought we could use [gitpython](https://github.com/gitpython-developers/GitPython), but even though it looks very widely used it's actually now in maintainance mode, which isn't great.

TomNicholas replied:
Tracking this in #30

TomNicholas commented on cstar_ocean/cstar_ocean/utils.py:
```diff
@@ -0,0 +1,149 @@
+
+def _replace_text_in_file(file_path, old_text, new_text):
+    """
+    Find and replace a string in a text file.
```
could probably just compile a regex instead

TomNicholas commented on cstar_ocean/examples/cstar_example.py:
```diff
@@ -0,0 +1,25 @@
+    caseroot="roms_marbl_example_case/",
+)
+
+## In a python session, execute:
+# roms_marbl_case.setup()
+# roms_marbl_case.build()
+# roms_marbl_case.pre_run()
+# roms_marbl_case.run(account_key=None) #substituting your account key on any HPC system
+# roms_marbl_case.post_run()
```
This should form the basis of an automated integration test

dafyddstephenson replied:
Agreed.

TomNicholas commented on cstar_ocean/setup.py:
```diff
@@ -0,0 +1,25 @@
+    version="0.0.0",
+    description="Computation Systems for Tracking Ocean Carbon (CSTAR) python package",
+    author="Dafydd Stephenson, Thomas Nicholas, Matt Long, and others",
+    author_email="dafydd@cworthy.org, thomas@cworthy.org",
```
```suggestion
    author_email="dafydd@cworthy.org, tom@cworthy.org",
```

TomNicholas reviewed (commented):
This is a great effort! I'm also very pleased how you've managed to get lots of the suggestions from our discussions in-person into this already.

TomNicholas commented:
> Are you suggesting it's worth merging this in sooner and then addressing those sorts of things down the line?

Yes. I think 5-minute things you can just do here, but anything beyond that should be addressed in separate PRs.

NoraLoose commented:
Really impressive work @dafyddstephenson! I looked through the PR but didn't have any further immediate "5-minute comments" beyond the ones that Tom already made. 

I am also confused about the whole conda installation thing. Have you tried to simply install via
```
cd cstar_ocean
pip install -e .
```
?

I also wonder if we want to move everything from the subdirectory `cstar_ocean` to the main level repository. Maybe we can move all the other subdirectories like `setup_tools` into a separate repo.

dafyddstephenson commented:
Thanks @NoraLoose ! Yes, I'm going to follow Tom's suggestion and boot a conda install to a future PR. It was included here as I wanted to be sure from the get-go we could manage the environment for compiling all this stuff without burdening the user with figuring it out - conda can just install the netcdf/gfortran/mpich etc dependencies as part of the C-Star install and I worked on getting that locked in first before coding anything up. Now I've figured it out though it can wait a while

dafyddstephenson commented:
We were discussing in the all hands merging this in and then doing a deep clean restructure for the next PR. I think I've now addressed all the comments (realised I hadn't updated the README or example files, now handled, I've also added some more detail on the various C-Star objects in the example notebook following feedback from @TomNicholas & @matt-long ). Will wait for a green light from everybody before hitting merge, but I think @NoraLoose is waiting on the merge so she can start working on some docs (thanks Nora!)

NoraLoose commented:
Green light from me 👍

dafyddstephenson commented:
Thanks for translating the comments into issues Tom - ready to go on the merge?

TomNicholas commented:
I think so yes!
